constraints/miden-vm/hash.air

```diff
@@ -0,0 +1,243 @@
+    # over to the nex row.
+    enf is_unchanged([s[1]]) when !f_out & !f_out_next
+    enf is_unchanged([s[2]]) when !f_out & !f_out_next
+
+    # Enforce that if any of f_abp, f_mpa, f_mva, f_mua flags is set to 1, the next value of s[0] 
+    # is 0.
+    enf s[0]' * f_comp = 0
+
+    # Enforce that no invalid combinations of flags are allowed.
+    enf cycle_row_7 * binary_not(s[0]) * s[1] = 0
+
+
+# Enforce node index constraints
+ev node_index(main: [s[3], i]):
+    # We can define two flags:
+    # 1. Flag f_hout = cycle_row_7 * binary_not(s[0]) * binary_not(s[1]) * binary_not(s[2]), 
+    # which is set to 1 when selector flags are (0,0,0) on rows which are 1 less than a multiple 
+    # of 8. This is flag of the instruction that returns the result of the currently running 
+    # computation.
+    # 2. Flag f_sout = cycle_row_7 * binary_not(s[0]) * binary_not(s[1]) * s[2], which is set to 1 
+    # when selector flags are (0,0,1) on rows which are 1 less than a multiple of 8. This is flag 
+    # of the instruction that returns the whole hasher state.
+    #
+    # Flag f_out is set to 1 when either f_hout = 1 or f_sout = 1 in the current row.
+    let f_out = cycle_row_7 * binary_not(s[0]) * binary_not(s[1])
+
+    # Set to 1 when selector flags are (1,0,1) on rows which are multiples of 8. This is flag of 
+    # the instruction that initiates Merkle path verification computation.
+    let f_mp = get_f_mp(s)
+
+    # Set to 1 when selector flags are (1,1,0) on rows which are multiples of 8. This is flag of 
+    # the instruction that initiates Merkle path verification for the "old" node value during 
+    # Merkle root update computation.
+    let f_mv = get_f_mv(s)
+
+    # Set to 1 when selector flags are (1,1,1) on rows which are multiples of 8. This is flag of 
+    # the instruction that initiates Merkle path verification for the "new" node value during
+    # Merkle root update computation.
+    let f_mu = get_f_mu(s)
+
+    # Set to 1 when selector flags are (1,0,1) on rows which are 1 less than a multiple of 8. This 
+    # is flag of the instruction that absorbs the next Merkle path node into the hasher state 
+    # during Merkle path verification computation.
+    let f_mpa = get_f_mpa(s)
+
+    # Set to 1 when selector flags are (1,1,0) on rows which are 1 less than a multiple of 8. This 
+    # is flag of the instruction that absorbs the next Merkle path node into the hasher state 
+    # during Merkle path verification for the "old" node value during Merkle root update 
+    # computation.
+    let f_mva = get_f_mva(s)
+
+    # Set to 1 when selector flags are (1,1,1) on rows which are 1 less than a multiple of 8. This 
+    # is flag of the instruction that absorbs the next Merkle path node into the hasher state 
+    # during Merkle path verification for the "new" node value during Merkle root update 
+    # computation.
+    let f_mua = get_f_mua(s)
```

Comment:
Similar to my previous comment - let's put all of these comments above the function definitions and remove them from here. That way this can be just 6 lines and this evaluator will be much easier to read. Let's also do that anywhere else any of these `get_f_*` functions are called